File: 백준/Gold/14938. 서강그라운드/서강그라운드.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.Arrays;
import java.lang.Math;

public class Main {
    static final int INF = 10_000_000;
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int r = Integer.parseInt(st.nextToken());
        
        int[] items = new int[n + 1];
        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= n; i++) {
            items[i] = Integer.parseInt(st.nextToken());
        }
        
        int[][] graph = new int[n + 1][n + 1];
        for (int i = 1; i <= n; i++) {
            Arrays.fill(graph[i], INF);
            graph[i][i] = 0;
        }
        
        for (int i = 0; i < r; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            graph[a][b] = Math.min(graph[a][b], c);
            graph[b][a] = Math.min(graph[b][a], c);
        }
        
        for (int k = 1; k <= n; k++) {
            for (int a = 1; a <= n; a++) {
                for (int b = 1; b <=n; b++) {
                    graph[a][b] = Math.min(graph[a][b], graph[a][k] + graph[k][b]);
                }
            }
        }
        
        int max_items = 0;
        for (int i = 1; i <= n; i++) {
            int total_items = 0;
            for (int j = 1; j <= n; j++) {
                if (graph[i][j] <= m) {
                    total_items += items[j];
                }
            }
            max_items = Math.max(max_items, total_items);
        }
        System.out.println(max_items);
    }
}

File: 백준/Gold/14938. 서강그라운드/서강그라운드.py
import sys
input = sys.stdin.readline
INF = int(1e9)
    
def main():
    n, m, r = map(int, input().split())
    items = list(map(int, input().split()))
    graph = [[INF] * (n + 1) for _ in range(n + 1)]
    
    for i in range(1, n + 1):
        graph[i][i] = 0

    for _ in range(r):
        a, b, c = map(int, input().split())
        graph[a][b] = min(graph[a][b], c)
        graph[b][a] = min(graph[b][a], c)
        
    for k in range(1, n + 1):
        for a in range(1, n + 1):
            for b in range(1, n + 1):
                graph[a][b] = min(graph[a][b], graph[a][k] + graph[k][b])
                
    max_items = 0
    for i in range(1, n + 1):
        total_items = 0
        for j in range(1, n + 1):
            if graph[i][j] <= m:
                total_items += items[j - 1]
        max_items = max(max_items, total_items)
     
    print(max_items)

if __name__ == "__main__":
    main()